feat(donate):add donation modal and certification message

ahmaxed opened the pull request:
<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [X] My pull request targets the `master` branch of freeCodeCamp.
- [X] None of my changes are plagiarized from another source without proper attribution.
- [X] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [X] My changes do not use shortened URLs or affiliate links.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

ahmaxed committed:
feat: add donation modal on block completion and donation message on certificate

ahmaxed committed:
fix: make completion percentage use props

ahmaxed committed:
fix: wait for completion modal to close and one second to pass before opening donation modal.

ahmaxed committed:
fix: close Donation modal on /donate redirection

ahmaxed committed:
fix: hide certificate message for the session if user clicks on /donate link

ahmadabdolsaheb commented:
This pr will display a donation message on a certificate if the username of the certificate matches the username of the signed in user.

The message will disappear for the session if user clicks on the message link and get redirected to donate page.

Additionally a modal will be displayed after the last challenge  of a block has been submitted.

if either of the donation requests has been shown during a session, the other one will not be shown.

Please let me know if the logic is in place so I could proceed with tests.

ahmadabdolsaheb commented:
@ojeytonwilliams, could you take a quick look?

raisedadead commented:
Please provide detailed context about when this modal is to be shown for QA.

ahmadabdolsaheb commented:
@raisedadead, the modal will be shown after the last challenge on a block has been completed.
it will check if the user has donated, if any other form of donation has been requested (certificate message) in that session.

an easy way to test it would be the data viz certificates. the donation should pop up after the last challenge has been completed.

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -33,13 +36,27 @@ export function* currentChallengeSaga({ payload: id }) {
+  let shouldRequestDonation = yield select(shouldRequestDonationSelector);
+  if (shouldRequestDonation) {
+    yield take(challTypes.closeModal);
+    yield delay(1000);
```
I think this could break if the user has a slow connection and takes longer than a second to mount the next page.

I'm not sure what the perfect solution is, though. I just feel like the donation modal should be triggered by the mounting of a new page, rather than the closing of another modal.

ahmadabdolsaheb replied:
Yes, you are right. previously it used to wait for challengeMounted action, but that cannot be used now since the next page after the last challenge in a block is an into and challangeMounted will not be called. Maybe we should call a introMounted action for the intros and listen to that for this donation modal. what do you think?

ahmadabdolsaheb replied:
note to self, there are times when the user has finished the last challenges but the last remaining challenge is in the middle of the block meaning closing the completion modal will be followed by challenge Mounted.

ojeytonwilliams replied:
How about putting it in the Learn layout?  That way, no matter where the user ends up after completing a challenge, as long an they're still in /learn they'll see the modal.  I think that might be more robust than checking in challenges and intros.

ojeytonwilliams commented on client/src/client-only-routes/ShowCertification.js:
```diff
@@ -103,8 +124,31 @@ class ShowCertification extends Component {
+
+    let conditionalDonationMessage = '';
+
+    if (signedInUserName === username && !isDonating && !isDonationRequested) {
```
This only displays if you navigate to a cert internally.  If you reload a cert it doesn't appear.  Is that intended?

ahmadabdolsaheb replied:
hmm I see, should I fetch user then?

ojeytonwilliams replied:
Not sure.  I think you need to do something similar to `ShowSettings`:

https://github.com/freeCodeCamp/freeCodeCamp/blob/3304b95213a44525ab2912c2e9987cd49013dc63/client/src/client-only-routes/ShowSettings.js#L167-L170

since `isSignedIn` is false when the page is first loaded.  I don't have much time today, or I'd look into this a bit closer.  Sorry! Should be able to tomorrow.

ojeytonwilliams reviewed (changes requested):
The UX is great, but I'm not sure if it's always behaving exactly as it's supposed to and there's potentially an issue for slow connections:

ahmaxed committed:
Update client/src/client-only-routes/ShowCertification.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams committed:
WIP: trigger donation modal on all learn pages

ojeytonwilliams committed:
fix: only show modal if block has been finished

ojeytonwilliams committed:
fix: show donate message when cert is reloaded

Tries to fetch the user if they appear not to be signed in and logs the
pageview with ga.

ahmaxed committed:
fix: change styles and add heart svg

ahmaxed committed:
feat: add heart anim

ahmaxed committed:
feat: add example test

ahmaxed committed:
feat: adjust syles

ahmadabdolsaheb commented:
@ojeytonwilliams could you take another look?

ojeytonwilliams commented on client/src/redux/index.js:
```diff
@@ -53,12 +55,14 @@ const initialState = {
   [
     'appMount',
+    'hardGoTo',
+    'canRequestDonation',
```
```suggestion
    'allowDonationRequests',
```
I named this poorly.

ojeytonwilliams commented on client/src/redux/index.js:
```diff
@@ -85,6 +90,10 @@ export const sagas = [
+export const tryToShowDonationModal = createAction(
+  types.tryToShowDonationModal
+);
+export const canRequestDonation = createAction(types.canRequestDonation);
```
```suggestion
export const allowDonationRequests = createAction(types.allowDonationRequests);
```

ojeytonwilliams commented on client/src/redux/index.js:
```diff
@@ -209,6 +212,10 @@ function spreadThePayloadOnUser(state, payload) {
 
 export const reducer = handleActions(
   {
+    [types.canRequestDonation]: state => ({
```
```suggestion
    [types.allowDonationRequests]: state => ({
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -4,7 +4,8 @@ import store from 'store';
   updateComplete,
-  updateFailed
+  updateFailed,
+  canRequestDonation
```
```suggestion
  allowDonationRequests
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -33,13 +34,18 @@ export function* currentChallengeSaga({ payload: id }) {
   yield put(updateSuccessMessage(randomCompliment()));
 }
 
+export function* canRequestDonationSaga() {
```
```suggestion
export function* allowDonationRequestsSaga() {
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -33,13 +34,18 @@ export function* currentChallengeSaga({ payload: id }) {
 }
 
+export function* canRequestDonationSaga() {
+  yield put(canRequestDonation());
```
```suggestion
  yield put(allowDonationRequests());
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -33,13 +34,18 @@ export function* currentChallengeSaga({ payload: id }) {
     takeEvery(types.challengeMounted, currentChallengeSaga),
-    takeEvery(types.challengeMounted, updateSuccessMessageSaga)
+    takeEvery(types.challengeMounted, updateSuccessMessageSaga),
+    takeEvery(types.lastBlockChalSubmitted, canRequestDonationSaga)
```
```suggestion
    takeEvery(types.lastBlockChalSubmitted, allowDonationRequestsSaga)
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.test.js:
```diff
@@ -0,0 +1,12 @@
+/* global  expect */
+import { canRequestDonationSaga } from './current-challenge-saga';
```
```suggestion
import { allowDonationRequestsSaga } from './current-challenge-saga';
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.test.js:
```diff
@@ -0,0 +1,12 @@
+import { canRequestDonationSaga } from './current-challenge-saga';
+import { types as appTypes } from '../../../redux';
+
+describe('canRequestDonationSaga', () => {
```
```suggestion
describe('allowDonationRequestsSaga', () => {
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.test.js:
```diff
@@ -0,0 +1,12 @@
+import { types as appTypes } from '../../../redux';
+
+describe('canRequestDonationSaga', () => {
+  it('should call canRequestDonation', () => {
```
```suggestion
  it('should call allowDonationRequests', () => {
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.test.js:
```diff
@@ -0,0 +1,12 @@
+
+describe('canRequestDonationSaga', () => {
+  it('should call canRequestDonation', () => {
+    const gen = canRequestDonationSaga();
```
```suggestion
    const gen = allowDonationRequestsSaga();
```

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.test.js:
```diff
@@ -0,0 +1,12 @@
+  it('should call canRequestDonation', () => {
+    const gen = canRequestDonationSaga();
+    expect(gen.next().value.payload.action.type).toEqual(
+      appTypes.canRequestDonation
```
```suggestion
      appTypes.allowDonationRequests
```

ojeytonwilliams reviewed (changes requested):
I noticed I'd named an action badly, so I added some suggestions to fix that.  I didn't change `donationRequested` to `preventDonationRequests`, because I was worried I'd miss one.  I can add that as a separate commit.

As far as functionality goes, everything looks good.  The following is just nitpicking about names:

ahmaxed committed:
Update client/src/assets/icons/Heart.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/assets/icons/Heart.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/assets/icons/Heart.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/assets/icons/Heart.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/components/Donation/components/DonationModal.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.test.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.test.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.test.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.test.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.test.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/components/Donation/Donation.css

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/components/Donation/Donation.css

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/redux/index.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/redux/index.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/redux/index.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/templates/Challenges/redux/current-challenge-saga.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ojeytonwilliams committed:
refactor: make action name more descriptive

ojeytonwilliams commented on client/src/components/Donation/components/DonationModal.js:
```diff
@@ -0,0 +1,102 @@
+};
+
+class DonateModal extends Component {
+  constructor(...props) {
```
```suggestion
```

ojeytonwilliams commented on client/src/components/Donation/components/DonationModal.js:
```diff
@@ -0,0 +1,102 @@
+
+class DonateModal extends Component {
+  constructor(...props) {
+    super(...props);
```
```suggestion
```

ojeytonwilliams commented on client/src/components/Donation/components/DonationModal.js:
```diff
@@ -0,0 +1,102 @@
+class DonateModal extends Component {
+  constructor(...props) {
+    super(...props);
+  }
```
```suggestion
```

ojeytonwilliams commented on client/src/components/Donation/components/DonationModal.js:
```diff
@@ -0,0 +1,102 @@
+  constructor(...props) {
+    super(...props);
+  }
+
```
```suggestion
```

ojeytonwilliams reviewed (commented):
Eslint's complaining about the constructor:

ahmaxed committed:
Update client/src/components/Donation/components/DonationModal.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/components/Donation/components/DonationModal.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/components/Donation/components/DonationModal.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

ahmaxed committed:
Update client/src/components/Donation/components/DonationModal.js

Co-Authored-By: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>
